docs/design/fabric-v2+/fpc-registration.puml

```diff
@@ -32,197 +35,203 @@ note over ERCC1, ERCC2 #FFaa00
+    CC_Id, version, MSPID, and channel id can be retrieved through Fabric
+    shim functions; TLCC_mrenclave can be retrieved through ERCC;
+
+    Note that the validation of these parameters is performed by ERCC.
+    If necessary, a pre-validation can be performed through the Ledger Enclave.
+    Another alternative could be to pass these parameters as argument in
+    the createEnclave call.
   end note
-  Enclave1 -> Enclave1: fetch CC_Parameters
 
-  note over Enclave1 #FFaa00
-    This step generates the enclave's private keys used by the FPC chaincode
-    to sign statements and set up secure channels.
-    Check out the Components diagram to visualize Public_Keys/Private_Keys data structures.
+  Enclave1 -> Enclave1 : <Public_Keys, Private_Keys>  <- generate keys
+  note right Enclave1
+    This step generates the enclave's public and private keys used by
+    the FPC chaincode to sign transactions and decrypt transaction arguments.
+    See fpc-components.uml diagram for more details
   end note
-  Enclave1 -> Enclave1 : Private_Keys <- generate_keys()
-  Enclave1 -> Enclave1 : Public_Keys <- Public(Private_Keys)
 
   Enclave1 -> Enclave1 : Attested_Data <- <CC_Parameters, Public_Keys>
 
-group "enclave attestation: Evidence <- GetEvidence(Attested_Data)"
-  note over E_Manager #FFaa00
-    As the EPID attestation requires the SPID, we let the E_Manager provide it.
-    In a concrete implementation, the E_Manager could explicitly send the SPID to ECC_Enclave.
-    Also, the SPID could be made an optional argument within an Aux_Data field.
-    This would allow for the generalization of the attestation, to enable DCAP.
+  Enclave1 -> Enclave1 +: GetEvidence(Attested_Data)
+  group enclave attestation - evidence creation
+  note right Enclave1
+    The GetEvidence method abstracts/generalizes the attestation protocol used here.
+    Currently, Intel SGX supports two different attestation schemes, namely,
+    EPID-based attestation and DCAP. For more details see Intel SGX docu.
+
+    In this diagram we illustrate the flow of the EPID-based attestation, which
+    involves Intel Attestation Service (IAS). As the EPID attestation requires
+    the SPID to generate a quote, the peer provides it through FPC_stub.
   end note
-  Enclave1-> E_Manager ++: getSPID
+  Enclave1-> Enclave1 ++: getSPID
   return SPID
 
-  Enclave1 -> Enclave1 : Attestation <- attest(SPID, Attested_Data)
-  note left #FFaa00
-    quote is defined with reportbody = Hash(<Attested_Data>)
+  Enclave1 -> Enclave1 : quote <- create quote using SPID over Hash(Attested_Data)
+  note right Enclave1
+      There are differnt options to implmement getSPID:
+      - FPC_stub could explicitly send the SPID to FPC_stub_enclave
+      - SPID could be an optional argument within an Aux_Data field.
+      - FPC_stub_enclave performs a ocall to fetch the SPID from FPC_stub
   end note
 
-  note over Enclave1 #FFaa00
+  Enclave1 -> IAS ++: send quote for verification
```

Comment:
(just a comment -- though we had talked and agreed about this)
On a second thought, this message to IAS looks more like an implementation detail than an abstraction/generalization. Also, since it originates from the enclave/chaincode, who's familiar with Fabric will most certainly be confused. Anyway, let's look for feedback.

Reply:
> since it originates from the enclave/chaincode,

well, it's the chaincode _stub_, not the (fpc_)chaincode itself. So it wouldn't really be a contradiction?  (Note since externalbuilder, we do not really have a golang CC anymore where the golang chaincode is the stub for fpc)

Reply:
But that's still chaincode-side logic requesting data from a remote service. And if you implement it through an invoke, it must be a "single query".

Regarding the external builder/golang cc, I guess you mean "we do have it and use it; we could avoid it, but it requires a major effort to provide similar/same support."

Reply:
> But that's still chaincode-side logic requesting data from a remote service. And if you implement it through an invoke, it must be a "single query".

we don't mention `invoke`, here, though, so the fact that we might initially implement that way shouldn't be confusing? 
Regarding invoke vs query: with the new flows we anyway want `query` rather than `invoke` not only for the "single" aspect you mention but also as we don't really need/want a transaction on the create --- that's what the register is for -- and it side-steps any issues like the one we just run into this morning (see #381 ) where the invoke for `getSPID` caused an (unnecessary) error. The invoke, though, would implicitly ensure that only a a proper designated peer could create an enclave, so there would be some plus: i.e., the easiest way to implement consistency with designated peer policy ....  :-)
 
> Regarding the external builder/golang cc, I guess you mean "we do have it and use it; we could avoid it, but it requires a major effort to provide similar/same support."

Well, we use golang stub and have functions which look like a (peculiar) golang chaincode, but that's all essentially invisible to the fabric core (as it is for the viewer of this diagram).  Which actually seems exactly the right way?  I.e., we do not show the particular implementation, which also is not really pertinent to understand the flows, and it leaves the freedom to do whatever is easiest now -- and i definitely was not arguing for moving away the current fpc-stub = go-stub + fpc-loader/bridge-code-ooking-like-go-chaincode! -- but leaves all freedom to change it if it ever would be useful.